### PR TITLE
rt: add new api for oneshot op submission and creation

### PR DESCRIPTION
**Pull request by Noah-Kennedy**
This change adds a new API for creating and submitting oneshot operations. This new API is intended to obsolete the existing system, however transferring over has been left out of this PR to keep the change small.

It is intended that a similar API for multishot operations be landed in a followup PR as well. This was also left out of this PR to keep the change small.

This refactoring paves the way for further work on SQE linking, multishot operations, and other improvements and additions.

The goal of this change is to allow us to split up the oneshot and multishot logic to allow for a cleaner and more extensible system, allow for user-provided operations, and allow users to control when and how their ops get submitted to the squeue.

### CODE REVIEW

File: src/io/socket.rs
Comment (FrankReh) [inline src/io/socket.rs:58]: That's interesting.

File: src/runtime/driver/op/mod.rs
Comment (FrankReh) [inline src/runtime/driver/op/mod.rs:40]: Maybe add that a Future is being returned by this function?

File: src/runtime/driver/mod.rs
Comment (FrankReh) [inline src/runtime/driver/mod.rs:132]: Maybe say what unsafe contract is being upheld.

File: src/runtime/driver/mod.rs
Comment (FrankReh) [inline src/runtime/driver/mod.rs:125]: I think this function should be labelled `unsafe` and each ancestor becomes `unsafe` too until reaching `submit()` and there the rationale given for being safe is the unsafe contract is maintained by building the  squeue Entry per lifetime requirements. So then the unsafe chain can be seen going all the way back into each op. 

I see so many bugs and useless questions raised because people just cut-n-paste unsafe calls without ever reading into them and then writing their own responses at the call site.

File: src/io/write.rs
Comment (FrankReh) [inline src/io/write.rs:38]: I would prefer to see this impl that starts at line 38 which is responsible for the struct being init and the squeue Entry being created appearing right after the struct, at line 18. That would make this file better mirror the layout it used to have.

File: examples/mix.rs
Comment (FrankReh) [inline examples/mix.rs:44]: Would InfoFuture have avoided these changes being required?

**Review by FrankReh — APPROVED** [review]
Nice. Very nice how you kept the old mechanism working while introducing the new one next to it.

File: src/runtime/driver/op/mod.rs
Comment (Noah-Kennedy) [reply src/runtime/driver/op/mod.rs:49]: This helps set the stage for adding public handles.

File: src/runtime/driver/mod.rs
Comment (Noah-Kennedy) [reply src/runtime/driver/mod.rs:132]: This was copied from the earlier stuff. When I do a cleanup PR later when I've moved over all the ops, I can add this.

File: src/runtime/driver/mod.rs
Comment (Noah-Kennedy) [reply src/runtime/driver/mod.rs:125]: I agree

File: examples/mix.rs
Comment (Noah-Kennedy) [reply examples/mix.rs:44]: Generally yes

**Comment by ollie-etl** [comment]
@Noah-Kennedy I feel somewhat put out having not begin requested for input here, given the history (especially #165, #171, #120, #218). 

In addition, this PR appears to have had a lifespan of 8 hours, with minimal discussion, and all _other_ Pr's on this subject have met a (deserved) wall of "let us discuss". It would appear there is a very distinct weighting of opinions here, which I think is a shame. Its certainly discouraging for future involvement.

**Comment by FrankReh** [comment]
@ollie-etl It is understandable you feel that way but it is just a difference of working for Noah at the moment. Your input is very valuable. We hadn't heard from you in a while *and* this is just one of several PRs Noah wants to get in *and* he's very amenable to changing things quickly as we see fit. What's going in is by no means the final. It's just a for him and I to make progress while we work on things in parallel for a few days. Please don't feel put out.

**Comment by FrankReh** [comment]
@ollie-etl I'll add for myself, another reason I often push back harder on others than on Noah in cases like this is I feel bad if I let someone's PR go through and then Noah or I go about changing it in a few days anyway. It muddies the `git blame` output and doesn't give full credit to the person who did all the work going in.

**Comment by FrankReh** [comment]
@ollie-etl Also, hopefully you noticed I didn't wait for Noah to address my questions with the first PR.

And coincidentally, I'm getting my old multi-shot work from before your submissions to line with up at least your recent submissions, just so it compiles again and has a test case working. Knowing I will have to change it again when Noah as the multi version of the new API ready.

**Comment by FrankReh** [comment]
@ollie-etl Not getting an official review request just means the next step for us isn't gated by a third party. The GitHub Notifications page shows anyone watching this when things are being proposed.

When I haven't totally agreed with you or others, I have meant to step out of the way and let @Noah-Kennedy form the second opinion for approval. In a project like this; it's just not  possible to wait for three or more confirmations, there are too many ways of getting something good done. That doesn't mean three or more reviews aren't welcome. Multiple reviews are very welcome. There is a lot to learn from each other - obviously.

**Comment by FrankReh** [comment]
> @FrankReh I get notification, but 8 hours is quite a turn around time.

Yes. But I would have made it even shorter if I had had the energy when the PR was first put in.

I realize, probably Noah does too, that later PRs are harder to fully comment on because the page doesn't show all the diffs. We'll have to work around that somehow. We can't let GitHub's javascript, as great as it is, totally dictate things for us.

I suspect it's not too late to comment to still make comments on this PR that Noah and I will see.

**Comment by FrankReh** [comment]
@ollie-etl One more thing. All that being said. Is there a double standard? Well yes. Noah is the tokio core team member. Ultimately things go the way he wants. If you and I don't like it, there are other options. But you have to admit, he is one of the most welcoming people on the tokio team and when he has time, he is very thoughtful in his responses. So he's not a benevolent dictator on the project, but he could be if he wanted to be. The project is only here because he kept it alive for well over a year by himself.

**Comment by ollie-etl** [comment]
@FrankReh

- We're broadly in agreement that this change is a good thing :tm: 

- Pull requests for things which are useful to me have languished for months, with endless requests for discussion and changes, which I've mostly accommodated, albeit with gritted teeth at times. I cannot argue that process hasn't made the code better / introduced things which I hadn't considered.

- This is a PR into an area with known _existing_ discussion. There is not even a mention of prior art or why this approach is favored in the PR or in the feedback.

- 8 hours for a new feature probably isn't enough time, for a project with contributors in multiple timezones. Bug fixes are obviously somewhat different. I have little interest is _also_ monitoring chat forums, in addition to github notifications.

- The question now, which I'm going to have to consider very hard for the next week or so is: given its very apparent the merit of a feature is weighted by contributor, and the significant time it takes to get Pr's into a state where they can be merged here (for me), would I just be better off forking / reevaluating one of the alternatives / starting afresh / or just accepting the significant time required to contribute features I see as required to this project? :shrug:

**Comment by FrankReh** [comment]
@ollie-etl That's very understandable from your side. I don't begrudge you that if you have a better option you should take it - especially because you've said you're doing this for your work and your company is paying you to make wise choices - but I hope you don't come to that decision. And I agree you've had some PRs that take a long time. And anything that had an endless discussion in it was because I was on the other end, I know that - if Noah had ever thought I was making too fine a point on something, he would have chimed in (was my feeling). In my own defense, I think if I wasn't at that end, you might not have heard anything for a while.

Again, ironically, in between this thread, I am literally copying, modifying your excellent work on the MultiCQEFuture to get a MultiCQEStream working. Knowing it is better than what I had cobbled together back in Sept and also knowing when it's done, it will live in one or two  branches just long enough to then go through another change.

You can also take your time to decide what to do. If you care to review things in the meantime anyway, that would be great.

Stepping back even further for a moment. This is a very hard open  source project. It marries Linux  io_uring with Rust and async Rust and tokio async Rust - which also makes a very interesting intersection to be at (and why I'm here, even without a job at the moment). There are a ton of ways to get things done in Rust and async Rust. I look at `hyper` and there is a single maintainer and they move fast and get a lot of feedback from users but they don't wait for all users to agree. And `axum` which I think has essentially two maintainers where there are lots of PRs submitted by others that can't all be accepted.

I am not trying to have the last word. Your last comment stood very well on it's own. You would be missed if you leave.

**Comment by ileixe** [comment]
Hi guys and @Noah-Kennedy.

Is there any following works or further discussion to introduce link ops here based on this API? (I tried to find but nothing found on my side.)

If not, we're willing to continue follow works for link ops on our own and want to hear what's design choices, idea and gaps in here.

**Pull request merged**
